### PR TITLE
Make useTranslationIfExists take an i18n object as a parameter

### LINKED ISSUE #6352: Fix naming of map tools bug appearing as translation strings
image

Related to #6340  to fix map attribution labels.

### PR DESCRIPTION
**Pull request by tephenavies**
### What this PR does

Change useTranslationIfExists to ensure that React components that use it to render translation strings use a i18next-react method of re-rendering when the language is changed

And add unit tests for useTranslationIfExists

Fixes #6352 #6336

### Checklist

-   [x] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
-   [ ] I've updated relevant documentation in `doc/`.
-   [ ] I've updated CHANGES.md with what I changed.

### CODE REVIEW

File: lib/Core/injectTerms.ts
Comment (tephenavies) [inline lib/Core/injectTerms.ts]: Probably should document somewhere that this file should also be improved to not import global i18next

File: lib/Language/languageHelpers.ts
Comment (zoran995) [inline lib/Language/languageHelpers.ts]: can it happen that `keyOrString` be undefined (specified by user)?

Comment (tephenavies) [reply lib/Language/languageHelpers.ts]: It shouldn't, since the type signature doesn't include `undefined`, but the function was being called wrongly by a .jsx file (`lib/ReactViews/Map/Panels/HelpPanel/HelpPanelItem.jsx `) with a parameter that was sometimes undefined. That's also been fixed in this PR

File: lib/ReactViews/Map/Panels/HelpPanel/HelpPanelItem.jsx
Comment (zoran995) [inline lib/ReactViews/Map/Panels/HelpPanel/HelpPanelItem.jsx]: this should actually check if `placeholderImage` is defined

File: lib/Language/languageHelpers.ts
Comment (zoran995) [reply lib/Language/languageHelpers.ts]: I just run a check on `trainerItems` if the title is undefined it will throw an error although it is typescript file
```json5
"trainerItems": [
  {
    //"title": "stepbystepguide.gettingstarted.title", // this will throw rror
    "footnote": "stepbystepguide.gettingstarted.footnote",
    "steps": [
      {
        //"title": "stepbystepguide.gettingstarted.step1.title", // this will throw error when opening steps guide
        "markdownDescription": "stepbystepguide.gettingstarted.step1.markdownDescription"
      }
    ]
  }
]
```

File: lib/Core/injectTerms.ts
Comment (staffordsmith83) [reply lib/Core/injectTerms.ts]: Added a TODO comment, but could be resolved before merging if easy?

File: lib/ReactViews/Map/Panels/HelpPanel/HelpPanelItem.jsx
Comment (staffordsmith83) [reply lib/ReactViews/Map/Panels/HelpPanel/HelpPanelItem.jsx]: Done

File: lib/Language/languageHelpers.ts
Comment (staffordsmith83) [reply lib/Language/languageHelpers.ts]: @steve9164 which way do you want to go on this?

**Review by nf-s — CHANGES_REQUESTED** [review]
Got an error when using `Tour` functionality

image

Steps:
- http://ci.terria.io/utif-i18n-parm/
- Change language to French
- Go to step by step guide
  - image
- Press "commence"
  - image

File: lib/Language/languageHelpers.ts
Comment (tephenavies) [reply lib/Language/languageHelpers.ts]: Probably there's an implicit `any` in the code that deals with `trainerItems` - anything that deals with JSON. You could fix it by explicitly checking existence of each property of the JSON object. Using `unknown` helps out here - e.g. `fetchJson(...) as unknown` and then `if (o.title) {...}`

**Comment by tephenavies** [comment]
The error occurred because a couple of calls to `applyTranslationIfExists` didn't have the i18n parameter. This should now be fixed for all uses of the function.

File: lib/Language/languageHelpers.ts
Comment (staffordsmith83) [reply lib/Language/languageHelpers.ts]: Resolved:
```ts
export function applyTranslationIfExists(
  keyOrString: unknown,
  i18n: i18n,
  options?: { [key: string]: string }
): string {
  // keyOrString could be undefined in some cases even if we type it as string
  if (isJsonString(keyOrString)) {
    if (keyOrString.indexOf(TRANSLATE_KEY_PREFIX) === 0) {
      const translationKey = keyOrString.substr(TRANSLATE_KEY_PREFIX.length);
      return i18n.exists(translationKey)
        ? i18n.t(translationKey, options)
        : translationKey;
    } else {
      return keyOrString;
    }
  }
  // Return an empty string if keyOrString was undefined. All functions that call this require a string to be returned
  return "";
}
```

Or should I still require keyOrString to be a string so TS can check what we pass it in most cases?

**Comment by staffordsmith83** [comment]
@nf-s think this is OK? https://github.com/TerriaJS/terriajs/pull/5706#discussion_r909294321

File: lib/Language/languageHelpers.ts
Comment (staffordsmith83) [reply lib/Language/languageHelpers.ts]: Perhaps this is better:

```ts
export function applyTranslationIfExists(
  keyOrString: string,
  i18n: i18n,
  options?: { [key: string]: string }
): string {
  // keyOrString could be undefined in some cases even if we type it as string
  if (isJsonString(keyOrString as unknown)) {
    if (keyOrString.indexOf(TRANSLATE_KEY_PREFIX) === 0) {
      const translationKey = keyOrString.substr(TRANSLATE_KEY_PREFIX.length);
      return i18n.exists(translationKey)
        ? i18n.t(translationKey, options)
        : translationKey;
    } else {
      return keyOrString;
    }
  }
  // Return an empty string if keyOrString was undefined. All functions that call this require a string to be returned
  return "";
}
```

Comment (staffordsmith83) [reply lib/Language/languageHelpers.ts]: Resolved

**Review by staffordsmith83 — APPROVED** [review]

**Review by staffordsmith83 — APPROVED** [review]

**Review by staffordsmith83 — APPROVED** [review]

**Review by staffordsmith83 — APPROVED** [review]

**Comment by staffordsmith83** [comment]
Oops on those repeated approvals, process got a bit mixed up with crispy initiating the PR and me taking over! Waiting for @nf-s to review final changes

File: lib/Language/languageHelpers.ts
Comment (nf-s) [inline lib/Language/languageHelpers.ts:23]: I would type `keyOrString` as `unknown` if you're using `isJsonString`

File: lib/ReactViews/Map/Panels/HelpPanel/HelpPanelItem.jsx
Comment (nf-s) [inline lib/ReactViews/Map/Panels/HelpPanel/HelpPanelItem.jsx]: Can these comments be removed?

File: lib/ReactViews/Map/TrainerBar/TrainerBar.tsx
Comment (nf-s) [inline lib/ReactViews/Map/TrainerBar/TrainerBar.tsx]: This can be removed

File: lib/ReactViews/Tools/Tool.tsx
Comment (nf-s) [inline lib/ReactViews/Tools/Tool.tsx]: Instead of removing the `applyTranslationifExists` - you can import global `i18next` and use `applyTranslationIfExists(this.props.toolName, i18next)`

**Review by nf-s — CHANGES_REQUESTED** [review]

File: lib/Language/languageHelpers.ts
Comment (zoran995) [reply lib/Language/languageHelpers.ts:23]: @nf-s I am more into this change as it will force us to use string type while and non-string value can come only from config and in the future values in config will be part of the traits system and validated as a string, just my thoughts on this

File: lib/ReactViews/Map/TrainerBar/TrainerBar.tsx
Comment (staffordsmith83) [reply lib/ReactViews/Map/TrainerBar/TrainerBar.tsx]: Done

File: lib/ReactViews/Tools/Tool.tsx
Comment (staffordsmith83) [reply lib/ReactViews/Tools/Tool.tsx]: @nf-s the issue here is that applyTranslationifExists will only translate anything that has "translate#" at the start of the string otherwise it will just return the passed in value. So its not using applyTranslationIfExists in the right way. Should I still leave it in if its doing nothing?

File: lib/ReactViews/Map/Panels/HelpPanel/HelpPanelItem.jsx
Comment (staffordsmith83) [reply lib/ReactViews/Map/Panels/HelpPanel/HelpPanelItem.jsx]: Yes, not needed any more, resolved.

File: lib/Language/languageHelpers.ts
Comment (staffordsmith83) [reply lib/Language/languageHelpers.ts:23]: Yes that what I was thinking @zoran995, if you see my previous commit I typed it as unknown, but then thought better to type as string so we are forced to pass a string in ost occurrences, and if a non-string value sneaks through we handle it inside `applyTranslationIfExists()`. Happy to defer to you @nf-s but do you think that intention is valid? Perhaps my type assertion in `if (isJsonString(keyOrString as unknown))` is not the correct way to do things.

**Review by nf-s — APPROVED** [review]

File: lib/ReactViews/Tools/Tool.tsx
Comment (nf-s) [reply lib/ReactViews/Tools/Tool.tsx]: Ok, I wasn't aware that tool name/title couldn't get string which started with `translate#`.

File: lib/Language/languageHelpers.ts
Comment (nf-s) [reply lib/Language/languageHelpers.ts:23]: Ok happy to leave it then

File: lib/ReactViews/Tools/Tool.tsx
Comment (staffordsmith83) [reply lib/ReactViews/Tools/Tool.tsx]: Nah you might be right @nf-s  Ill just check with @na9da about his intention for this part he had some ideas I think he wrote initially

File: lib/Language/languageHelpers.ts
Comment (na9da) [reply lib/Language/languageHelpers.ts:23]: In my latest debugging, I think `name` and `title` doesn't get used anywhere. Will clean it up sometime - https://github.com/TerriaJS/terriajs/issues/6368

File: lib/ReactViews/Tools/Tool.tsx
Comment (na9da) [reply lib/ReactViews/Tools/Tool.tsx]: In my latest debugging, I think `name` and `title` doesn't get used anywhere. Will clean it up sometime - https://github.com/TerriaJS/terriajs/issues/6368

**Comment by na9da** [comment]
I have done a few tests to see if it impacts other tools like `DiffTool`, but they seem to be working fine.

**Pull request merged**
